Source/WebCore/html/ValidatedFormListedElement.cpp

```diff
@@ -0,0 +1,494 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ValidatedFormListedElement.h"
+
+#include "ControlStates.h"
+#include "ElementAncestorIterator.h"
+#include "Event.h"
+#include "EventHandler.h"
+#include "EventNames.h"
+#include "FormAssociatedElement.h"
+#include "FormController.h"
+#include "Frame.h"
+#include "FrameView.h"
+#include "HTMLDataListElement.h"
+#include "HTMLFieldSetElement.h"
+#include "HTMLFormElement.h"
+#include "HTMLLegendElement.h"
+#include "HTMLParserIdioms.h"
+#include "PseudoClassChangeInvalidation.h"
+#include "ScriptDisallowedScope.h"
+#include "ValidationMessage.h"
+#include <wtf/Ref.h>
+#include <wtf/SetForScope.h>
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+using namespace HTMLNames;
+
+ValidatedFormListedElement::ValidatedFormListedElement(HTMLFormElement* form)
+    : FormListedElement(form)
+{
+}
+
+ValidatedFormListedElement::~ValidatedFormListedElement()
+{
+    clearForm();
+}
+
+bool ValidatedFormListedElement::willValidate() const
+{
+    if (!m_willValidateInitialized || m_dataListAncestorState == DataListAncestorState::Unknown) {
+        m_willValidateInitialized = true;
+        bool newWillValidate = computeWillValidate();
+        if (m_willValidate != newWillValidate)
+            m_willValidate = newWillValidate;
+    } else {
+        // If the following assertion fails, updateWillValidateAndValidity() is not
+        // called correctly when something which changes computeWillValidate() result
+        // is updated.
+        ASSERT(m_willValidate == computeWillValidate());
+    }
+    return m_willValidate;
+}
+
+bool ValidatedFormListedElement::computeWillValidate() const
+{
+    if (m_dataListAncestorState == DataListAncestorState::Unknown) {
+#if ENABLE(DATALIST_ELEMENT)
+        const HTMLElement& element = asHTMLElement();
+        m_dataListAncestorState = (element.document().hasDataListElements() && ancestorsOfType<HTMLDataListElement>(element).first()) ? DataListAncestorState::InsideDataList : DataListAncestorState::NotInsideDataList;
+#else
+        m_dataListAncestorState = DataListAncestorState::NotInsideDataList;
+#endif
+    }
+    // readonly bars constraint validation for *all* <input> elements, regardless of the <input> type, for compat reasons.
+    return m_dataListAncestorState == DataListAncestorState::NotInsideDataList && !isDisabled() && !m_hasReadOnlyAttribute;
+}
+
+void ValidatedFormListedElement::updateVisibleValidationMessage(RefPtr<HTMLElement> anchor)
+{
+    HTMLElement& element = asHTMLElement();
+    Page* page = element.document().page();
```

Comment:
It's oddly specific that we check nullity of Page but never use Page object...
Maybe add a FIXME here?

Reply:
I've removed `Page* page =` variable in [519a6cb](https://github.com/WebKit/WebKit/pull/4988/commits/519a6cbdfbc48c9f6df22e3d1ae44cd2cf8134bf), yet IMO the idea of is quite straightforward: don't attempt to show any message for detached frames.